llvm/include/llvm/Option/OptStrCmp.h

```diff
@@ -0,0 +1,51 @@
+//===- OptStrCmp.h - Option String Comparison -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_OPTION_OPTSTRCMP_H
+#define LLVM_OPTION_OPTSTRCMP_H
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace llvm::opt {
+
+// Comparison function for Option strings (option names & prefixes).
+// The ordering is *almost* case-insensitive lexicographic, with an exception.
+// '\0' comes at the end of the alphabet instead of the beginning (thus options
+// precede any other options which prefix them). Additionally, if two options
+// are identical ignoring case, they are ordered according to case sensitive
+// ordering if `FallbackCaseSensitive` is true.
+static int StrCmpOptionName(StringRef A, StringRef B,
+                            bool FallbackCaseSensitive = true) {
+  size_t MinSize = std::min(A.size(), B.size());
+  if (int Res = A.substr(0, MinSize).compare_insensitive(B.substr(0, MinSize)))
+    return Res;
+
+  // If they are identical ignoring case, use case sensitive ordering.
+  if (A.size() == B.size())
+    return FallbackCaseSensitive ? A.compare(B) : 0;
+
+  return (A.size() == MinSize) ? 1 /* A is a prefix of B. */
+                               : -1 /* B is a prefix of A */;
+}
+
+// Comparison function for Option prefixes.
+template <typename ContainerTy>
+static int StrCmpOptionPrefixes(const ContainerTy &APrefixes,
+                                const ContainerTy &BPrefixes) {

```

Comment:
Does this have to be a template? Couldn't we get away with something like this?
```c++
int StrCmpOptionPrefixes(ArrayRef<StringRef> APrefixes,
                         ArrayRef<StringRef> BPrefixes);
```

In `OptEmitter.cpp`, you should be able to pass `APrefixes` and `BPrefixes` straight into the function, since `std::vector<StringRef>` implicitly converts to `ArrayRef<StringRef>`. In `OptTable.cpp` you might need to do something like this:

```c++
StrCmpOptionPrefixes({A.Prefixes.data(), A.Prefixes.size()},
                     {B.Prefixes.data(), B.Prefixes.size()})
```

... which should trigger implicit conversion from `StringLiteral *` to `StringRef *`, call the `ArrayRef<StringLiteral>` constructor and be fine I think.

Reply:
I did try that earlier. Note that in general it's not safe to convert from `ArrayRef<SubClass>` to `ArrayRef<BaseClass>`. However, in this case its fine since StringLiteral does not add any new data to StringRef. I will do this with an added static assert to verify this.

Reply:
I implemented the suggestion. PTAL.

Reply:
Looks good!